pkg/security/security_profile/activity_tree/activity_tree.go

```diff
@@ -0,0 +1,378 @@
+	invalidRootNodeReason NodeDroppedReason = "invalid_root_node"
+	bindFamilyReason      NodeDroppedReason = "bind_family"
+	allDropReasons                          = []NodeDroppedReason{eventTypeReason, brokenLineageReason, invalidRootNodeReason, bindFamilyReason}
+)
+
+// NodeGenerationType is used to indicate if a node was generated by a runtime or snapshot event
+// IMPORTANT: IT MUST STAY IN SYNC WITH `adproto.GenerationType`
+type NodeGenerationType byte
+
+const (
+	// Unknown is a node that was added at an unknown time
+	Unknown NodeGenerationType = 0
+	// Runtime is a node that was added at runtime
+	Runtime NodeGenerationType = 1
+	// Snapshot is a node that was added during the snapshot
+	Snapshot NodeGenerationType = 2
+	// SecurityProfile is a node that was added because of a drift from a security profile
+	SecurityProfile NodeGenerationType = 3
+)
+
+func (genType NodeGenerationType) String() string {
+	switch genType {
+	case Runtime:
+		return "runtime"
+	case Snapshot:
+		return "snapshot"
+	default:
+		return "unknown"
+	}
+}
+
+// ActivityTreeOwner is used to communicate with the owner of the activity tree
+type ActivityTreeOwner interface {
+	MatchesSelector(entry *model.ProcessCacheEntry) bool
+	IsEventTypeValid(evtType model.EventType) bool
+	NewProcessNodeCallback(p *ProcessNode)
+}
+
+// ActivityTree contains a process tree and its activities. This structure has no locks.
+type ActivityTree struct {
+	Stats *ActivityTreeStats
+
+	treeType          string
+	shouldMergePaths  bool
+	differentiateArgs bool
+
+	validator ActivityTreeOwner
+
+	CookieToProcessNode map[uint32]*ProcessNode `json:"-"`
+	ProcessNodes        []*ProcessNode          `json:"-"`
+
+	// top level lists used to summarize the content of the tree
+	DNSNames     *utils.StringKeys
+	SyscallsMask map[int]int
+}
+
+// NewActivityTree returns a new ActivityTree instance
+func NewActivityTree(validator ActivityTreeOwner, treeType string) *ActivityTree {
+	at := &ActivityTree{
+		treeType:            treeType,
+		validator:           validator,
+		Stats:               NewActivityTreeNodeStats(),
+		CookieToProcessNode: make(map[uint32]*ProcessNode),
+		SyscallsMask:        make(map[int]int),
+		DNSNames:            utils.NewStringKeys(nil),
+	}
+	return at
+}
+
+// ComputeSyscallsList computes the top level list of syscalls
+func (at *ActivityTree) ComputeSyscallsList() []uint32 {
+	output := make([]uint32, 0, len(at.SyscallsMask))
+	for key := range at.SyscallsMask {
+		output = append(output, uint32(key))
+	}
+	return output
+}
+
+// IsEmpty returns true if the tree is empty
+func (at *ActivityTree) IsEmpty() bool {
+	return len(at.ProcessNodes) == 0
+}
+
+// nolint: unused
+func (at *ActivityTree) debug(w io.Writer) {
+	for _, root := range at.ProcessNodes {
+		root.debug(w, "")
+	}
+}
+
+// ScrubAndRetainProcessArgsEnvs scrubs and retains process args and envs
+func (at *ActivityTree) ScrubAndRetainProcessArgsEnvs(resolver *process.Resolver) {
```

Comment:
now should be `ScrubProcessArgsEnvs` . we don't need to retain anymore

Reply:
👍🏻